raft-tech/TANF-app
# Feat/1866 celery beat and postgres backups

## Summary of Changes
_Provide a brief summary of changes_
Pull request closes #1866 

## How to Test

1. Deploy to any dev environment.
2. Reset env `cf set-env tdp-backend-<name> REDIS_URI "redis://localhost:6379"
3. `cf restage tdp-backend-<name>` to get this value picked up otherwise celery beat won't connect to redis.
4. Watch app log where in s3 it is storing the pg dump.
5. You can additionally verify the s3 location manually via:
```
cf ssh tdp-backend-<name>
cd app
/tmp/lifecycle/shell
python manage.py shell_plus
import boto3, os, sys, json
In [5]: 
   ...: 
   ...: OS_ENV = os.environ

In [7]: sys_values = {}

In [8]: sys_values['S3_ENV_VARS'] = json.loads(OS_ENV['VCAP_SERVICES'])['s3']


In [10]: sys_values['S3_CREDENTIALS'] = sys_values['S3_ENV_VARS'][0]['credentials']
    ...: sys_values['S3_URI'] = sys_values['S3_CREDENTIALS']['uri']
    ...: sys_values['S3_ACCESS_KEY_ID'] = sys_values['S3_CREDENTIALS']['access_key_id']
    ...: 
    ...: sys_values['S3_SECRET_ACCESS_KEY'] = sys_values['S3_CREDENTIALS']['secret_acce
    ...: ss_key']
    ...: sys_values['S3_BUCKET'] = sys_values['S3_CREDENTIALS']['bucket']
    ...: sys_values['S3_REGION'] = sys_values['S3_CREDENTIALS']['region']
    ...: sys_values['DATABASE_URI'] = OS_ENV['DATABASE_URL']

In [11]: def list_s3_files(sys_values):
    ...:     """List the files in s3 bucket."""
    ...:     """
    ...:     :param bucket:
    ...:     :param region:
    ...:     """
    ...:     session = boto3.Session(aws_access_key_id=sys_values['S3_ACCESS_KEY_ID'],
    ...:                             aws_secret_access_key=sys_values['S3_SECRET_ACCESS
    ...: _KEY'],
    ...:                             region_name=sys_values['S3_REGION'])
    ...:     s3 = session.resource('s3')
    ...:     my_bucket = s3.Bucket(sys_values['S3_BUCKET'])
    ...:     return list(_ for _ in my_bucket.objects.all())
    ...: 

In [12]: list_s3_files(sys_values)
Out[12]: 
[s3.ObjectSummary(bucket_name='cg-178858c2-2794-44ac-a18a-c8f6efe4197a', key='/backup./test_backup.pg'),
 s3.ObjectSummary(bucket_name='cg-178858c2-2794-44ac-a18a-c8f6efe4197a', key='/backup/home/vcap/test.pg'),
 s3.ObjectSummary(bucket_name='cg-178858c2-2794-44ac-a18a-c8f6efe4197a', key='/backup/test.txt'),
 s3.ObjectSummary(bucket_name='cg-178858c2-2794-44ac-a18a-c8f6efe4197a', key='/backup/tmp/5-14-2022-01.pg'),
 s3.ObjectSummary(bucket_name='cg-178858c2-2794-44ac-a18a-c8f6efe4197a', key='/backup/tmp/backup-test.pg'),
 s3.ObjectSummary(bucket_name='cg-178858c2-2794-44ac-a18a-c8f6efe4197a', key='/backup/tmp/backup.pg'),
 s3.ObjectSummary(bucket_name='cg-178858c2-2794-44ac-a18a-c8f6efe4197a', key='/backup/tmp/backup1.pg'),
 s3.ObjectSummary(bucket_name='cg-178858c2-2794-44ac-a18a-c8f6efe4197a', key='/backup/tmp/backup_test.pg'),
 s3.ObjectSummary(bucket_name='cg-178858c2-2794-44ac-a18a-c8f6efe4197a', key='/backup/tmp/file1.pg'),
 s3.ObjectSummary(bucket_name='cg-178858c2-2794-44ac-a18a-c8f6efe4197a', key='/backup/tmp/test.pg'),
 s3.ObjectSummary(bucket_name='cg-178858c2-2794-44ac-a18a-c8f6efe4197a', key='/backup/tmp/test_backup_ticket.pg'),
```

> *Demo GIF(s) and screenshots for testing procedure*
![1866](https://user-images.githubusercontent.com/84722778/197215776-6d887a70-99a0-4ae6-8320-ac54fb55f2cd.JPG)

## Deliverables
_More details on how deliverables herein are assessed included [here](../docs/How-We-Work/our-priorities-values-expectations.md#Deliverables)._

### [Deliverable 1: Accepted Features](../docs/How-We-Work/our-priorities-values-expectations.md#Deliverable-1-Accepted-Features)

Checklist of ACs:
- [x] Each deployed environment will have nightly backups of the postgres database
- [x] Backups are obsolesced after 30 days
- [x] Local dev environments do not backup
- [x] Data backups at rest are encrypted and secure

### [Deliverable 2: Tested Code](../docs/How-We-Work/our-priorities-values-expectations.md#Deliverable-2-Tested-Code)

+ Are all areas of code introduced in this PR meaningfully tested?
+ Are code coverage minimums met?

### [Deliverable 3: Properly Styled Code](../docs/How-We-Work/our-priorities-values-expectations.md#Deliverable-3-Properly-Styled-Code)


### [Deliverable 4: Accessible](../docs/How-We-Work/our-priorities-values-expectations.md#Deliverable-4-Accessibility)



### [Deliverable 5: Deployed](../docs/How-We-Work/our-priorities-values-expectations.md#Deliverable-5-Deployed)


### [Deliverable 6: Documented](../docs/How-We-Work/our-priorities-values-expectations.md#Deliverable-6-Code-documentation)


### [Deliverable 7: Secure](../docs/How-We-Work/our-priorities-values-expectations.md#Deliverable-7-Secure)


### [Deliverable 8: User Research](../docs/How-We-Work/our-priorities-values-expectations.md#Deliverable-8-User-Research)

Research product(s) clearly articulate(s):
